test/Test_TypeAssignment.re

```diff
@@ -0,0 +1,603 @@
+        ])
+      }
+    ),
+  );
+
+//Generates a UExp with no regard to the typing rules
+let pure_random_uexp: QCheck.Gen.t(UExp.t) =
+  QCheck.Gen.sized_size(
+    int_range(1, 5),
+    fix((self, n) =>
+      switch (n) {
+      | 0 =>
+        frequency([
+          (1, map(uexp_int, int)),
+          (1, map(uexp_bool, bool)),
+          (1, map(uexp_float, float)),
+          (1, map(uexp_string, printable_string)),
+          (1, map(uexp_var, printable_string)),
+        ])
+      | n =>
+        frequency([
+          (1, map3(uexp_binop, int_int_binop, self(n / 2), self(n / 2))),
+          (1, map3(uexp_binop, int_bool_binop, self(n / 2), self(n / 2))),
+          (
+            1,
+            map3(uexp_binop, float_float_binop, self(n / 2), self(n / 2)),
+          ),
+          (
+            1,
+            map3(uexp_binop, float_bool_binop, self(n / 2), self(n / 2)),
+          ),
+          (1, map3(uexp_binop, bool_bool_binop, self(n / 2), self(n / 2))),
+          (
+            1,
+            map3(uexp_fun, printable_string, pure_random_utyp, self(n / 2)),
+          ),
+          (1, map(uexp_list, list_size(int_range(1, 5), self(n / 2)))),
+          (1, map2(uexp_ap, self(n / 2), self(n / 2))),
+          (1, map3(uexp_if, self(n / 2), self(n / 2), self(n / 2))),
+          (
+            1,
+            map2(
+              uexp_match,
+              self(n / 2),
+              list_size(
+                int_range(1, 4),
+                pair(pure_random_upat, self(n / 2)),
+              ),
+            ),
+          ),
+        ])
+      }
+    ),
+  );
+
+//Generates alcotest test cases
+let rec testcases_gen =
+        (size: int, ty: UTyp.t, l: list(test_case(unit)))
+        : list(test_case(unit)) => {
+  switch (size) {
+  | 0 => l
+  | n =>
+    let u = uexp_gen([], ty, 1);
+    let m = Test_Elaboration.mk_map(u);
+    let d = Elaborator.dhexp_of_uexp(m, u, false);
+    let test = () =>
+      Alcotest.check(
+        Alcotest.pass,
+        "Random expression",
+        unit,
+        TypeAssignment.property_test(Some(Term.UTyp.to_typ([], ty)), d, m)
+          ? unit : Alcotest.fail("Term: " ++ UExp.show(u)),
+      );
+    let case = test_case("Type assignment", `Quick, test);
+    testcases_gen(n - 1, ty, [case, ...l]);
+  };
+};
+
+let type_assignment_tests = testcases_gen(10, utyp(Int), []);
+
+//Generate purely random uexp test cases for alcotest
+let random_tests =
+  List.map(
+    (u: UExp.t) => {
+      let m = Test_Elaboration.mk_map(u);
+      let d = Elaborator.dhexp_of_uexp(m, u, false);
+      let ty = Elaborator.fixed_exp_typ(m, u);
+      let test = () =>
+        Alcotest.check(
+          Alcotest.pass,
+          "Random expression",
+          unit,
+          TypeAssignment.property_test(ty, d, m)
+            ? unit : Alcotest.fail("Term: " ++ UExp.show(u)),
+        );
+      test_case("Type assignment", `Quick, test);
+    },
+    QCheck.Gen.generate(~n=10, pure_random_uexp),
```

Comment:
I think we probably want each property to be it's own test as opposed to each generated value resulting in a new test. We could then mark the tests as `Slow` instead of `Quick` so we don't slow the test runs down. It's worth looking into https://github.com/c-cube/qcheck?tab=readme-ov-file#integration-within-alcotest to see if we can go straight from the qcheck runner to alcotest.

Reply:
It's also worth doing something on the assertion to be closer to the equality we want to test so the error messages are better. We can probably do better but this is something:

```reason
let testable_typ = testable(Fmt.using(Typ.pretty_print, Fmt.string), Typ.eq);

//Generate purely random uexp test cases for alcotest
let random_tests =
  List.map(
    (u: UExp.t) => {
      print_endline("Term: " ++ UExp.show(u));
      let m: Statics.Map.t = Test_Elaboration.mk_map(u);
      let d: option(DHExp.t) = Elaborator.dhexp_of_uexp(m, u, false);
      let ty: option(Typ.t) = Elaborator.fixed_exp_typ(m, u);
      let dhexp_typ =
        Option.bind(d, TypeAssignment.typ_of_dhexp(Builtins.ctx_init, m));
      let test = () =>
        Alcotest.check(
          Alcotest.option(testable_typ),
          "Random expression: " ++ UExp.show(u),
          dhexp_typ,
          ty,
        );
      test_case("Type assignment", `Quick, test);
    },
    QCheck.Gen.generate(~n=10, pure_random_uexp),
  );
```

Reply:
Qcheck has first class support for generating long/short tests https://github.com/c-cube/qcheck?tab=readme-ov-file#long-tests